### PR TITLE
Initial artwork submission form

### PR DESCRIPTION
**Pull request by jobcurtis**
#234

Here's a draft PR for the artwork submission form. It's branched from #235 so includes commits from that PR as well (that PR should be merged before this one). It's not complete yet. The main things missing still are;

- [x] validation / storage of PD proof fields
- [x] allow specifying that artwork completed year is approximate
- [x] creation of thumbnails for uploads
- [ ] general form UX - we definitely need some help text / placeholder values to guide users a bit, and the layout could probably be improved.

### CODE REVIEW

**Comment by jobcurtis** [comment]
Hey @colagrosso, I think this is ready for a second pair of eyes. This PR contains;

- required fields for submitting a new artwork
- basic validation
- checking that uploaded images are actually images, storing them & creating a thumbnail

Things that aren't in this PR but should be or could be included later;

- more helpful feedback when failing validation
- previously filled-in fields should be set when you're redirected to the form after hitting an error
- more validation around the pd-proof fields, particularly the museum URL option
- guidance text / placeholder values

Also note that apache is currently configured to reject requests where the body is > 2MB -- I think we will need to add an exception to that rule for the `/artworks/post` route.

File: www/artworks/post.php
Comment (colagrosso) [inline www/artworks/post.php:26]: It's worth taking a moment to ensure we like this naming convention. We end up with files that look like this:

```
$ ls -1 www/images/uploads/
02584f01-fc89-4dec-aeff-9db5f8b97055.jpeg
02584f01-fc89-4dec-aeff-9db5f8b97055.thumb.jpeg
0ee63a18-eda7-4289-9a95-d5e6742ec08e.jpeg
0ee63a18-eda7-4289-9a95-d5e6742ec08e.thumb.jpeg
361a0483-940f-4520-b09a-d36f89c937c0.jpeg
361a0483-940f-4520-b09a-d36f89c937c0.thumb.jpeg
97063843-0b29-48d8-a215-14898fbf1dff.jpeg
97063843-0b29-48d8-a215-14898fbf1dff.thumb.jpeg
c3803beb-a3b6-4e18-977b-8c3ab6b9bec8.jpeg
c3803beb-a3b6-4e18-977b-8c3ab6b9bec8.thumb.jpeg
cfa67358-40eb-44d2-abef-5b67bbea1e77.jpeg
cfa67358-40eb-44d2-abef-5b67bbea1e77.thumb.jpeg
ea2b1c23-df11-49ef-bdc7-00d0a17ca353.jpeg
ea2b1c23-df11-49ef-bdc7-00d0a17ca353.thumb.jpeg```

Alternatives

1. A `thumbs/` subdirectory
2. A different separator, e.g., `_thumb.jpeg

The reason I ask is that we'll need to add a new property to the `Artwork` class:

```
_ThumbnailUrl
```

that is derived from `ImageFilesystemPath`. We might want to make it easy on ourselves to derive the thumbnail URLs. Maybe the dot will work, but we'll have to careful with the string or regex logic.

File: lib/Constants.php
Comment (colagrosso) [inline lib/Constants.php:38]: I think we could bump this up to 350.

File: www/artworks/post.php
Comment (colagrosso) [inline www/artworks/post.php:43]: I think we should accept only JPEG uploads. It would simplify this code. Plus, the eventual ebook producer will need to supply a `cover.source.jpg`, so a PNG or GIF will be less useful.

File: www/artworks/post.php
Comment (colagrosso) [inline www/artworks/post.php:112]: This leads to database fields that look like this:

```
/standardebooks.org/web/www/images/uploads/c01d7ac5-09e2-4e32-ab3f-985dab0f7d2b.jpeg
```

which I think is not quite right. Let's not write the `WEB_ROOT` to the database, just something like this:

```
/images/uploads/c01d7ac5-09e2-4e32-ab3f-985dab0f7d2b.jpeg
```

That way, it will also serve as a download URL.

File: www/artworks/new.php
Comment (colagrosso) [inline www/artworks/new.php:84]: I think you noted this, too, but we should give some guidance to the form submitter here. Maybe as a short paragraph, maybe as some tooltips. (We can do this in a subsequent PR.)

File: README.md
Comment (jobcurtis) [reply README.md:19]: A `.gitignore` that ignores everything in the directory (except itself) with an explanatory comment should do the trick. Will add.

File: www/artworks/new.php
Comment (jobcurtis) [reply www/artworks/new.php:49]: I agree that it needs refinement - what if we eventually have works by 10,000 artists, are we going to include them all in the form html? Hopefully not.

I suppose the most straightforward way to achieve autofill for the death year field would be an accompanying JS script that retrieves that data from a `data-death-year` attribute on the artist in the datalist. We would want to get Alex's OK on that first though as there is no JS elsewhere in the site. In this case it might be alright as it wouldn't be providing critical functionality -- users with scripts disabled would just be missing autofill for one field, which isn't the end of the world.

File: lib/Artwork.php
Comment (colagrosso) [reply lib/Artwork.php:88]: Ah, sorry, you're right of course. My mistake.

File: www/artworks/new.php
Comment (colagrosso) [reply www/artworks/new.php:49]: It's not an issue of scale, just convenience and correctness. 

Let me think about this some more and maybe prototype something. I was imagining a form where the user would either pick an existing artist (both name and death year), or enter both manually. What we have now is a mix where the user picks the name but must always enter the death year manually.

No Javascript is a [core principle](https://alexcabal.com/posts/standard-ebooks-and-classic-web-tech), so I would prefer not to add it here.

File: www/artworks/post.php
Comment (acabal) [reply www/artworks/post.php:26]: GUIDs are not very ergonomic in the case where we have to go in and examine or manipulate files by hand. Instead we can simply name the file after the DB primary key value (i.e. `1.jpg`, `2.jpg`, etc.). That way we also avoid needing another column for GUID. Also let's go with `.jpg` and not `.jpeg`.

**Comment by acabal** [comment]
Let me expand on what I wrote about JS a little because this fuzzy-match textbox is a good example.

We should design the system to work without JS in any case, because the user may have JS disabled and they should still be able to use the system. However, it's clear that often JS enhances a UI. So we can add JS later as an *enhancement*, but the underlying system should still work without it.

The datalist for artist names is a good example of this. `<datalist>` may not be good enough to fuzzy match alternate spellings and accent marks. So the system should probably prompt the user after POST to disambiguate those issues, before actually recording the form data:

1. User enters `Ilia Repyn` (an alternate spelling) into the form and submits
2. POST function checks the DB for variant spellings and sees the user probably meant `Ilya Repin`. 
3. Instead of showing confirmation page, user is returned to the form with a messages saying `Did you mean Ilya Repin?`
4. User can select the alternate spelling we suggest, or submit using original name

This is a flow without JS, but that would be clearly enhanced with a fully-featured fuzzy match JS textbox, because we could easily disambiguate before the form is submitted. So we can explore adding such a fuzzy JS textbox later, but the important point is that the flow works without JS anyway.

File: www/artworks/post.php
Comment (jobcurtis) [reply www/artworks/post.php:26]: If the filename is going derived from the PK we can remove the column that stores the file location, right? As the file location will always be `/www/images/uploads/{id}.jpg`.

File: README.md
Comment (jobcurtis) [reply README.md:20]: This was a guess on my part as I'm not running Ubuntu so my apache setup is a bit different. I'll remove this and make a note on the issue that we need to figure out the correct steps for the readme.

File: www/artworks/new.php
Comment (jobcurtis) [reply www/artworks/new.php:84]: Agreed, let's leave this for later once the functionality of the form is complete.

File: lib/Artist.php
Comment (jobcurtis) [reply lib/Artist.php:46]: I'd prefer to be explicit about what's being selected (and in what order) unless there's a good reason not to. These tables aren't necessarily set in stone and `SELECT *` is prone to quietly break if the table definition changes.

File: www/artworks/post.php
Comment (jobcurtis) [reply www/artworks/post.php:74]: Good catch, I've fixed this (and the issue with the empty tags)

File: lib/Artist.php
Comment (colagrosso) [reply lib/Artist.php:46]: I understand your reasoning, but I think consistency with the codebase is just as important. There's a pretty consistent use of `SELECT *`, including for PropertiesBase objects: 

https://github.com/search?q=repo%3Astandardebooks%2Fweb%20%22select%20*%22&type=code

File: www/artworks/post.php
Comment (colagrosso) [reply www/artworks/post.php:35]: @acabal, how do feel about the full-resolution image URLs being guessable? Going with `1.jpg`, `2.jpg`, etc. would allow potentially annoying scrapers:

```
for i in $(seq 1 500); do
  curl -O https://standardebooks.org/www/images/uploads/$i.jpg
done
```

(This came up in a [thread](https://groups.google.com/g/standardebooks/c/KrXyzZVZig4/m/E7BRuJ2wBQAJ) in Mar–Apr 2021 about bulk downloading ebooks.)

If you don't think that's too big of an issue, or it's still worth the trade-off, then I agree with you and Job that we can simplify here and remove the `ImageFilesystemPath` column. We'll still add these to properties to the `Artwork` class,

```
_Url
_ThumbnailUrl
```

but they will be trivial.

Comment (acabal) [reply www/artworks/post.php:35]: Having them guessable is fine. Unlikely to be scraped I think, nobody is seriously scraping hi-res public domain oil paintings without any accompanying metadata. If it becomes a problem we can address it at that time.

File: www/artworks/post.php
Comment (jobcurtis) [reply www/artworks/post.php]: Good point, I've done that.

File: www/artworks/post.php
Comment (jobcurtis) [reply www/artworks/post.php]: Good catch, I've fixed this so that `GetImageUrl()` returns e.g. `/images/uploads/1.jpg`.

File: www/artworks/post.php
Comment (jobcurtis) [reply www/artworks/post.php]: Cool, I've implemented this change so that uploads are stored at `/images/uploads/{id}.jpg` and thumbs at `/images/uploads/{id}.thumb.jpg`.

Another potential risk of is that if upload URLs are predictable, it would be possible to use the SE website as a free image hosting service since even unapproved uploads will be available. This also seems pretty unlikely (and not too difficult to detect) but maybe worth being aware of.

File: www/artworks/new.php
Comment (jobcurtis) [reply www/artworks/new.php]: Good idea, I've added the message with the name of the artwork in it.

File: www/artworks/new.php
Comment (jobcurtis) [reply www/artworks/new.php:49]: Cool - let's make a note on the main issue to handle this better at some point down the road.

File: www/artworks/post.php
Comment (jobcurtis) [reply www/artworks/post.php]: This function now returns `void` as the image/thumb paths are generated by the `Artwork` class instead.

**Comment by jobcurtis** [comment]
Alright, I think I've addressed all the initial feedback and it's ready for another look (@colagrosso). I'll handle the merge conflict shortly.

Regarding the autofill/search for existing artists, I think let's leave that for a separate PR after we've decided how it should be implemented. For this PR I'm happy to leave in the current datalist or take it out, whatever we think is appropriate.

File: www/artworks/post.php
Comment (colagrosso) [inline www/artworks/post.php:16]: nit: single quotes

File: www/artworks/post.php
Comment (colagrosso) [inline www/artworks/post.php:104]: nit: single quotes

File: www/artworks/post.php
Comment (colagrosso) [inline www/artworks/post.php:106]: nit: single quotes

File: www/images/uploads/.gitignore
Comment (colagrosso) [inline www/images/uploads/.gitignore:8]: nit: I think this red circle is indicating that there's not a newline at the end of this file. If that's the case, please add one.

**Review by colagrosso — APPROVED** [review]
> Alright, I think I've addressed all the initial feedback and it's ready for another look (@colagrosso). I'll handle the merge conflict shortly.

Great job, Job! LGTM. I left a few nits before merging. Sorry I didn't see them before.

Removing the `GetArtist()` function from `Artwork.php` should be enough to handle the merge conflict.

> Regarding the autofill/search for existing artists, I think let's leave that for a separate PR after we've decided how it should be implemented. For this PR I'm happy to leave in the current datalist or take it out, whatever we think is appropriate.

Let's leave the datalist in add address it in a separate PR. You've done more than enough work here for one PR already. We'll add some details on #234 on what we learned here about artist help and any other issues. 

> Also, this PR has grown to contain quite a lot of commits. I'd normally squash it down to one or two before merging unless anyone would rather keep the entire history. Thoughts?

I appreciated all the small commits during the review because it was easy to see what changed, but now I think it would be better to squash it down to one or two before merging.

**Pull request merged**
